### PR TITLE
Halo regions

### LINKED ISSUE #45: Use multiple dispatch to control Poisson solver
[This `if`-statement](https://github.com/ali-ramadhan/Oceananigans.jl/blob/e1be1f0e3f067cdb4236eb5e7b962e0c772ce4d1/src/time_steppers.jl#L140) is not necessary, because the architecture can be known from the array types of the `CellField`s.

Specifically, [this line](https://github.com/ali-ramadhan/Oceananigans.jl/blob/e1be1f0e3f067cdb4236eb5e7b962e0c772ce4d1/src/spectral_solvers.jl#L389) can be changed to

```
function solve_poisson_3d_ppn_planned!(ssp::SpectralSolverParameters, g::RegularCartesianGrid, f::CellField{T}, ϕ::CellField{T}) where T<:CuArray

...

end
```

I think it would actually be preferable to dispatch on the type of the arrays in `RegularCartesianGrid`. However, it does not appear that the arrays representing the coordinates of the grid are parameterized in [the definition of `RegularCartesianGrid`](https://github.com/ali-ramadhan/Oceananigans.jl/blob/e1be1f0e3f067cdb4236eb5e7b962e0c772ce4d1/src/grids.jl#L12). Why is that? 

If the array type of the problem is part of `RegularCartesianGrid`, then dispatch can be used in place of `if`-statements all over the code, especially in `fields.jl`.

Finally, the term 'spectral solvers' is misleading here. The solver does not have spectral accuracy; it simply uses the FFT, which *happens* to be used to solve different problems that have a 'spectral' character.

### LINKED ISSUE #104: Implement halo regions.
Even for single-core and single-GPU simulations. This might actually speed things up as we won't have to use `incmod1` and `decmod1` everywhere #57.

### LINKED ISSUE #122: We can reduce memory usage in the Poisson solver by a factor of 2.
We don't need to store both `f` and `ϕ`. When converting the `f` Fourier coefficients to `ϕ` coefficients, we can just overwrite the `f` array.

This will be great because the Poisson solver requires `Complex{Float64}` so we free up quite a bit of memory.

We can then get rid of `fCC2` in `StepperTemporaryFields`, move `fCC1` somewhere else like `PoissonSolver` or `Model`, and I think we don't need `fC1` anyways. So we get rid of the `StepperTemporaryFields` struct, and end up storing **zero** intermediate calculations!

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR introduces _halo regions_ instead of using wrap-around `indmod1` and `decmod1` functions to account for horizontally periodic boundary conditions. It will allow us to more easily implement multiple wall-bounded dimensions, e.g. channel models #100. This is also a good first step towards distributed parallelization with MPI.

Resolves #45 
Resolves #104 
Resolves #122

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
cc @jm-c 

This PR now removes the `StepperTemporaryFields` struct as it's not needed anymore. We don't need to store δρ and the `Complex{Float64}` fields were moved to the `PoissonSolver` struct to play nicely with the halo regions which are implemented using an `OffsetArray`.

But feel free to bring them back in (i.e. keep them in your branch) if they're needed for LES closures.

File: src/time_steppers.jl [hunk 1]
```diff
@@ -111,6 +112,54 @@ function solve_for_pressure!(::GPU, model::Model)
     @launch device(GPU()) idct_permute!(model.grid, ϕ, model.pr.pNHS.data, threads=(Tx, Ty), blocks=(Bx, By, Bz))
 end
 
+function fill_halo_regions!(arch::Architecture, grid::Grid, fields...)
+    Ty = min(max_threads, Ny)
+    Tz = min(fld(max_threads, Ty), Nz)
+    By, Bz = cld(Ny, Ty), cld(Nz, Tz)
+    @launch device(arch) fill_halo_regions_x!(grid, fields..., threads=(1, Ty, Tz), blocks=(1, By, Nz))
```
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply src/time_steppers.jl]: Interesting, I've been doing it this way for a while and everything has worked. You mean like this?
```julia
@launch device(arch) threads=(1, Ty, Tz), blocks=(1, By, Nz) fill_halo_regions_x!(grid, fields...)
```

**Comment by ali-ramadhan** [comment]
This PR is functionally ready but I might not merge it yet for three reasons:
1. The halo regions do not play super well with 1D and 2D models yet. Ideally we would not use halo regions in the dimension of size 1, but that would require writing in some more code. For now we always have halo regions in the _x_ and _y_ directions. So merging this may break some of the 1D and 2D examples. Also, some (30 pass, 6 fail, 4 error) of the boundary condition tests fail as they're 1D models.
2. We should add tests for the `fill_halo_regions!` functions. Maybe they belong in `fields.jl` instead of `time_steppers.jl`.
3. The golden master tests fails for velocity fields only, and pass for tracer fields. The difference is in the 7th decimal place. After thinking about this, I think this makes sense because on `#master` we were incorrectly using wavenumbers/eigenvalues of type `Float32` when the FFTs were in `Float64`. I've fixed this now so that `kx²`, `ky²`, and `kz²` are all `Float64`. So the model is more accurate now and the golden master file should be updated. It also explains why it's only failing for velocities and why they differ in the 7th decimal place (`eps(Float32) = 1.1920929f-7`).

I could work on these but I have other priorities right now.

I will branch off this PR to work on the channel model (#100). Hopefully things won't get too finicky...

File: src/time_steppers.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (vchuravy) [reply src/time_steppers.jl]: Yeah that's how it was intended. https://github.com/vchuravy/GPUifyLoops.jl/blob/951ddadfecb47b5636129d45467316aa185b32e3/src/GPUifyLoops.jl#L41

It seems to work due to the fact that we process the kwargs funnily...

[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply src/time_steppers.jl]: Ah good to know. I'll switch my `@launch` calls before any weird errors pop out.

**Comment by ali-ramadhan** [comment]
Suggestion by @glwagner: `data(::Field)` that returns the interior/data points.

**Comment by glwagner** [comment]
@ali-ramadhan Happy to add that function to this PR once the merge conflicts are resolved.

**Comment by ali-ramadhan** [comment]
That would be great! I just added it here as a note to remember, probably won't work on this PR again for a few more weeks.

**Comment by ali-ramadhan** [comment]
Superseded by #250 and #253.
